src/Chainweb/Miner/POW.hs

```diff
@@ -319,42 +325,46 @@ mine
     => Proxy a
     -> BlockHeader
     -> Nonce
-    -> IO BlockHeader
-mine _ h nonce = BA.withByteArray initialTargetBytes $ \trgPtr -> do
-    !ctx <- hashMutableInit @a
-    bytes <- BA.copy initialBytes $ \buf ->
-        allocaBytes (powSize :: Int) $ \pow -> do
-
-            -- inner mining loop
-            --
-            -- We do 100000 hashes before we update the creation time.
-            --
-            let go 100000 !n = do
-
-                    -- update the block creation time
-                    ct <- getCurrentTimeIntegral
-                    injectTime ct buf
-                    go 0 n
-
-                go !i !n = do
-
-                    -- Compute POW hash for the nonce
-                    injectNonce n buf
-                    hash ctx buf pow
-
-                    -- check whether the nonce meets the target
-                    fastCheckTarget trgPtr (castPtr pow) >>= \case
-                        True -> return ()
-                        False -> go (succ i) (succ n)
-
-            -- Start inner mining loop
-            go (0 :: Int) nonce
-
-    -- On success: deserialize and return the new BlockHeader
-    runGet decodeBlockHeaderWithoutHash bytes
-
+    -> IO (T2 BlockHeader Word)
+mine _ h nonce = do
+    counter <- newIORef 0
+    res <- BA.withByteArray initialTargetBytes $ \trgPtr -> do
+        !ctx <- hashMutableInit @a
+        bytes <- BA.copy initialBytes $ \buf ->
+            allocaBytes (powSize :: Int) $ \pow -> do
+
+                -- inner mining loop
+                --
+                -- We do 100000 hashes before we update the creation time.
+                --
+                let go 100000 !n = do
+
+                        -- update the block creation time
+                        ct <- getCurrentTimeIntegral
+                        injectTime ct buf
+                        go 0 n
+
+                    go !i !n = do
+
+                        -- increment hash counter
+                        modifyIORef' counter (+ 1)
```

Comment:
I would move this out of this loop and increment in steps of 1000.

Reply:
Sure.

Reply:
oops, sorry, I meant 100,000. Essentially, just moving up into the case `go 100000 !n`, avoiding any non-essential overhead in the inner loop.